lib/hpke/src/userinfo_vc.cpp

```diff
@@ -0,0 +1,197 @@
+    { "ES384", Signature::get<Signature::ID::P384_SHA384>() },
+    { "ES512", Signature::get<Signature::ID::P521_SHA512>() },
+    { "Ed25519", Signature::get<Signature::ID::Ed25519>() },
+    { "Ed448", Signature::get<Signature::ID::Ed448>() },
+    { "RS256", Signature::get<Signature::ID::RSA_SHA256>() },
+    { "RS384", Signature::get<Signature::ID::RSA_SHA384>() },
+    { "RS512", Signature::get<Signature::ID::RSA_SHA512>() },
+  };
+
+  return alg_sig_map.at(alg);
+}
+
+std::chrono::system_clock::time_point
+epoch_time(int64_t seconds_since_epoch)
+{
+  const auto delta = std::chrono::seconds(seconds_since_epoch);
+  return std::chrono::system_clock::time_point(delta);
+}
+
+struct UserInfoVC::ParsedCredential
+{
+  // Header fields
+  const Signature& signature_algorithm; // `alg`
+  std::string key_id;         // `kid`
+
+  // Top-level Payload fields
+  std::string issuer;                               // `iss`
+  std::chrono::system_clock::time_point not_before; // `nbf`
+  std::chrono::system_clock::time_point not_after;  // `exp`
+
+  // Credential subject fields
+  std::map<std::string, std::string> credential_subject;
+  Signature::ID public_key_algorithm;
+  std::shared_ptr<Signature::PublicKey> public_key;
+
+  // Signature verification information
+  bytes to_be_signed;
+  bytes signature;
+
+  static std::shared_ptr<ParsedCredential> parse(const std::string& jwt)
+  {
+    // Split the JWT into its header, payload, and signature
+    const auto first_dot = jwt.find_first_of('.');
+    const auto last_dot = jwt.find_last_of('.');
+    if (first_dot == std::string::npos || last_dot == std::string::npos ||
+        first_dot == last_dot) {
+      throw std::runtime_error("malformed JWT; not enough '.' characters");
+    }
+
+    const auto header_b64 = jwt.substr(0, first_dot);
+    const auto payload_b64 = jwt.substr(first_dot, last_dot - first_dot);
+    const auto signature_b64 = jwt.substr(last_dot);
+
+    // Parse the components
+    const auto header = json::parse(from_base64url(header_b64));
+    const auto payload = json::parse(from_base64url(payload_b64));
+    const auto to_be_signed = from_ascii(header_b64 + "." + payload_b64);
+    const auto signature = from_base64url(signature_b64);
+
+    // Verify the VC parts
+    const auto vc = payload.at("vc");
+
+    static const std::string context = "https://www.w3.org/2018/credentials/v1";
+    if (vc.at("context") != context) {
+      throw std::runtime_error("malformed VC: incorrect context value");
+    }
+
+    static const auto type = std::vector<std::string>{
+      "VerifiableCredential",
+      "UserInfoCredential",
+    };
+    if (vc.at("type") != type) {
+      throw std::runtime_error("malformed VC: incorrect type value");
+    }
+
+    // Parse the subject public key
+    auto credential_subject = vc.at("credentialSubject");
+    const auto id = credential_subject.at("id").get<std::string>();
+    credential_subject.erase("id");
```

Comment:
curious, why are we erasing the id here ?

Reply:
Because the ID is "consumed" by transforming it into the public key.  I can leave it, though.  Not a big deal.